backend/backend.go

```diff
@@ -27,6 +29,22 @@ type Backend interface {
 	State() (state.State, error)
 }
 
+// MultiState is an interface that a backend can implement to allow changing
+// between named states depending on the configured environment.
+type MultiState interface {
```

Comment:
I'd prefer to make this required for all states (make it part of Backend).

For legacy remote state we can return an error in the actual `remote-state.Backend` implementation instead.

The reason being is that going forward I don't want this to be optional and introduce more optional interfaces. It is difficult now due to backwards compatibility but if we force all new remote implementations to support it and only error for legacy then we can slowly fix them over one by one as we convert to the new API.

I believe that should be fully backwards compatible and simplifies all the checking.